[SPARK-37391][SQL]JdbcConnectionProvider tells if it modifies security context

tdg5 opened the pull request:
Augments the JdbcConnectionProvider API such that a provider can indicate that it will need to modify the global security configuration when establishing a connection, and as such, if access to the global security configuration should be synchronized to prevent races.


### What changes were proposed in this pull request?
As suggested by @gaborgsomogyi [here](https://github.com/apache/spark/pull/29024/files#r755788709), augments the `JdbcConnectionProvider` API to include a `modifiesSecurityContext` method that can be used by `ConnectionProvider` to determine when `SecurityConfigurationLock.synchronized` is required to avoid race conditions when establishing a JDBC connection.


### Why are the changes needed?
Provides a path forward for working around a significant bottleneck introduced by synchronizing `SecurityConfigurationLock` every time a connection is established. The synchronization isn't always needed and it should be at the discretion of the `JdbcConnectionProvider` to determine when locking is necessary. See [SPARK-37391](https://issues.apache.org/jira/browse/SPARK-37391) or [this thread](https://github.com/apache/spark/pull/29024/files#r754441783).


### Does this PR introduce _any_ user-facing change?
Any existing implementations of `JdbcConnectionProvider` will need to add a definition of `modifiesSecurityContext`. I'm also open to adding a default implementation, but it seemed to me that requiring an explicit implementation of the method was preferable.

A drop-in implementation that would continue the existing behavior is:
```scala
override def modifiesSecurityContext(
  driver: Driver,
  options: Map[String, String]
): Boolean = true
```


### How was this patch tested?
Unit tests, but I also plan to run a real workflow once I get the initial thumbs up on this implementation.

tdg5 commented on sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcConnectionProvider.scala:
```diff
@@ -53,12 +53,25 @@ abstract class JdbcConnectionProvider {
+   * @return True if the connection provider will need to modify the security configuration when
+   * initiating a connection with the given driver with the given options.
+   */
+  def needsModifySecurityConfiguration(driver: Driver, options: Map[String, String]): Boolean
```
I'm open to other naming ideas.

tdg5 replied:
Also open to a default implementation, but seemed better to force implementers to explicitly implement.

gaborgsomogyi replied:
Maybe we can call it `modifiesSecurityContext`.

tdg5 commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/connection/BasicConnectionProvider.scala:
```diff
@@ -48,4 +48,12 @@ private[jdbc] class BasicConnectionProvider extends JdbcConnectionProvider with
+    options: Map[String, String]
+  ): Boolean = {
+    val jdbcOptions = new JDBCOptions(options)
+    jdbcOptions.keytab == null || jdbcOptions.principal == null
```
This could also just return `false`, but I'm kind of confused about what's going on here. [This doc](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/jdbc/README.md#:~:text=Built%2Din%20CPs%20added%20which%20support%20kerberos%20authentication%20using%20keytab%20and%20principal) suggests that built-in CPs support kerberos auth, but if I understand correctly that should require a `keytab` and ` principal`. I'm inclined to believe that this `JdbcConnectionProvider` doesn't actually support kerberos auth, but maybe I'm misunderstanding.

tdg5 replied:
Oops, just realized while driving that I need to invert this check. Well fix when I'm back at the keyboard.

tdg5 commented on sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/jdbc/connection/ConnectionProviderSuite.scala:
```diff
@@ -107,12 +124,50 @@ class ConnectionProviderSuite
+      override val name: String = "test1"
+      override def canHandle(driver: Driver, options: Map[String, String]): Boolean = true
+      override def getConnection(driver: Driver, options: Map[String, String]): Connection = {
+        assert(Thread.holdsLock(SecurityConfigurationLock))
```
Open to better ideas on testing that the lock was acquired or not. Could change it to an actual lock, but preferred to avoid the downstream repercussions.

AmplabJenkins commented:
Can one of the admins verify this patch?

tdg5 commented:
@gaborgsomogyi

tdg5 commented:
This is not a work in progress as far as I am concerned/aware, but I marked it as a WIP until @gaborgsomogyi can give it an initial thumbs up.

tdg5 commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/connection/BasicConnectionProvider.scala:
```diff
@@ -48,4 +48,12 @@ private[jdbc] class BasicConnectionProvider extends JdbcConnectionProvider with
+    options: Map[String, String]
+  ): Boolean = {
+    val jdbcOptions = new JDBCOptions(options)
+    jdbcOptions.keytab != null && jdbcOptions.principal != null
```
This could also just return `false`, but I'm kind of confused about what's going on here. [This doc](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/jdbc/README.md#:~:text=Built%2Din%20CPs%20added%20which%20support%20kerberos%20authentication%20using%20keytab%20and%20principal) suggests that built-in CPs support kerberos auth, but if I understand correctly that should require a `keytab` and `principal`. I'm inclined to believe that this `JdbcConnectionProvider` doesn't actually support kerberos auth, but maybe I'm misunderstanding.

gaborgsomogyi replied:
`BasicConnectionProvider` is the default unsecure connection provider so return false is good.

HyukjinKwon commented:
ok to test

SparkQA commented:
**[Test build #145734 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/145734/testReport)** for PR 34745 at commit [`8a9749c`](https://github.com/apache/spark/commit/8a9749c5947f5e94cacfc36ecee7dbb5a415bc6c).

SparkQA commented:
**[Test build #145734 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/145734/testReport)** for PR 34745 at commit [`8a9749c`](https://github.com/apache/spark/commit/8a9749c5947f5e94cacfc36ecee7dbb5a415bc6c).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/145734/

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/50204/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/50204/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/50204/

LuciferYang commented on sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/jdbc/connection/ConnectionProviderSuite.scala:
```diff
@@ -68,12 +69,20 @@ class ConnectionProviderSuite
       override def canHandle(driver: Driver, options: Map[String, String]): Boolean = true
       override def getConnection(driver: Driver, options: Map[String, String]): Connection =
         throw new RuntimeException()
+      override def needsModifySecurityConfiguration(
```
Should I write the default method to `JdbcConnectionProvider` (return false) and override it only where necessary

tdg5 replied:
I'm definitely open to this, but defer to the community on whether that is appropriate or not.

It's worth noting that a default of `false` seems like the right answer to me, but Spark versions 3.1+ implicitly have a default of `true`. 

I think a default of `false` is sensible, but it would change what is currently the default behavior.

gaborgsomogyi replied:
Bad idea to add default implementation(at least within the trait). Such case a user may think it's not important part of the provider and will work with the default. This is not true for all the secure providers.

AmplabJenkins commented:
Can one of the admins verify this patch?

tdg5 commented:
Hello friends! Anything I can do to help move this PR forward? Presently my ability to upgrade past spark 3.0.x is blocked by the change in functionality that this PR aims to improve.

gaborgsomogyi reviewed (commented):
I think the direction looks good but I've some suggestions.

tdg5 commented:
@gaborgsomogyi I've updated the PR to include your suggestions (really just a rename and a comment on why `BasicConnectionProvider` can always return `false`). Have another look when you have a chance, please.

SparkQA commented:
**[Test build #146237 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/146237/testReport)** for PR 34745 at commit [`5666f9e`](https://github.com/apache/spark/commit/5666f9ea8c4430bbd1adbf85f8a42e6d360384e6).

SparkQA commented:
**[Test build #146237 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/146237/testReport)** for PR 34745 at commit [`5666f9e`](https://github.com/apache/spark/commit/5666f9ea8c4430bbd1adbf85f8a42e6d360384e6).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/50711/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/146237/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/50711/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/50711/

gaborgsomogyi commented:
Have you exexuted the integration tests?

gaborgsomogyi commented:
+ plz fix mima cheks...

tdg5 commented:
Mima check has been fixed (assuming adding an exclude is the appropriate solution). Any idea if the kubernetes failures are something to worry about? I can't tell from the output if that's related to my change. Seems far removed from what I can tell, but it can be hard to tell. I've rebased from master, so perhaps that will help also.

SparkQA commented:
**[Test build #146293 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/146293/testReport)** for PR 34745 at commit [`adc4453`](https://github.com/apache/spark/commit/adc44530e06383df0c6bd082311dc4299aff8a46).

SparkQA commented:
**[Test build #146293 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/146293/testReport)** for PR 34745 at commit [`adc4453`](https://github.com/apache/spark/commit/adc44530e06383df0c6bd082311dc4299aff8a46).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/146293/

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/50765/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/50765/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/50765/

tdg5 commented:
I think the k8s integration tests are generally borked. If you happen to know that's not true, let me know and I can investigate further.
![image](https://user-images.githubusercontent.com/2316989/146441506-640fa577-7a40-4757-a71e-dc0dfc0b025f.png)

SparkQA commented:
**[Test build #146297 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/146297/testReport)** for PR 34745 at commit [`80e3d8e`](https://github.com/apache/spark/commit/80e3d8eb4693aba5a535a4291fce11b9c4def0a6).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/50769/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/50769/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/50769/

SparkQA commented:
**[Test build #146297 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/146297/testReport)** for PR 34745 at commit [`80e3d8e`](https://github.com/apache/spark/commit/80e3d8eb4693aba5a535a4291fce11b9c4def0a6).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/146297/

tdg5 commented:
Any trick to re-run tests? The spark test that failed in Jenkins passes for me locally. Maybe a transient failure. No rush, I'll force push something if I don't hear back.

SparkQA commented:
**[Test build #146304 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/146304/testReport)** for PR 34745 at commit [`503eab1`](https://github.com/apache/spark/commit/503eab1ffb4661a5f8dbbb3ebfb3f0a629ea4328).

HyukjinKwon commented:
retest this please

HyukjinKwon commented:
@tdg5 mind taking a look at https://github.com/apache/spark/pull/34745/checks?check_run_id=4554725898 and enable Github Actions in your forked repository? Apache Spark repo uses the GitHub Actions resources of forked repository to run the build.

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/50776/

SparkQA commented:
**[Test build #146306 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/146306/testReport)** for PR 34745 at commit [`503eab1`](https://github.com/apache/spark/commit/503eab1ffb4661a5f8dbbb3ebfb3f0a629ea4328).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/50778/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/50776/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/50776/

SparkQA commented:
**[Test build #146304 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/146304/testReport)** for PR 34745 at commit [`503eab1`](https://github.com/apache/spark/commit/503eab1ffb4661a5f8dbbb3ebfb3f0a629ea4328).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/50778/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/50778/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/146304/

tdg5 commented:
SparkR test failure doesn't seem to relate to my commit... Rebased from master to force another build.

SparkR test failure seems to have been an ephemeral thing?
```
* installing the package to build vignettes
* creating vignettes ...Warning in file(con, "r") :
  cannot open file '/tmp/RtmpTxszZ3/xshell185ef769eca3b6': No such file or directory
Error in file(con, "r") : cannot open the connection
Error in setwd(owd) : cannot change working directory
Execution halted
sh: 0: getcwd() failed: No such file or directory
```

SparkQA commented:
**[Test build #146313 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/146313/testReport)** for PR 34745 at commit [`4b324b5`](https://github.com/apache/spark/commit/4b324b5146e27521155afa4c4433c73fd00cfcd9).

SparkQA commented:
**[Test build #146306 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/146306/testReport)** for PR 34745 at commit [`503eab1`](https://github.com/apache/spark/commit/503eab1ffb4661a5f8dbbb3ebfb3f0a629ea4328).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/50785/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/146306/

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/50786/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/50785/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/50785/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/50786/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/50786/

SparkQA commented:
**[Test build #146313 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/146313/testReport)** for PR 34745 at commit [`4b324b5`](https://github.com/apache/spark/commit/4b324b5146e27521155afa4c4433c73fd00cfcd9).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/146313/

gaborgsomogyi commented:
Some of the tests are flaky but we need a green build to merge anything so retest is the way.

Other than that plz update the description to reflect the latest code.

tdg5 commented:
Description updated and another round of tests is underway.

SparkQA commented:
**[Test build #146359 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/146359/testReport)** for PR 34745 at commit [`3a2b619`](https://github.com/apache/spark/commit/3a2b619e19fb33285cb16c1e0022f9a58c294862).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/50833/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/50833/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/50833/

SparkQA commented:
**[Test build #146359 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/146359/testReport)** for PR 34745 at commit [`3a2b619`](https://github.com/apache/spark/commit/3a2b619e19fb33285cb16c1e0022f9a58c294862).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/146359/

tdg5 commented:
Wahoo! All tests passed on the last build. Well, not the k8s integration tests, but I don't get the impression those are blocking. 🎉

gaborgsomogyi commented:
There are several manual integration tests which cover this feature, have you executed them manually?

tdg5 commented:
> There are several manual integration tests which cover this feature, have you executed them manually?

No, I was unaware of those. Any guidance available for running those? I will look into it next time I'm at a keyboard.

tdg5 commented:
Something like this I'm guessing?
```
ENABLE_DOCKER_INTEGRATION_TESTS=1 \
POSTGRES_DOCKER_IMAGE_NAME=postgres:14.0 \
MYSQL_DOCKER_IMAGE_NAME=mysql:5.7.36 \
MSSQLSERVER_DOCKER_IMAGE_NAME=2019-CU13-ubuntu-20.04 \
MARIADB_DOCKER_IMAGE_NAME=mariadb:10.5.12 \
DB2_DOCKER_IMAGE_NAME=ibmcom/db2:11.5.6.0a \
ORACLE_DOCKER_IMAGE_NAME=gvenzl/oracle-xe:latest \
./build/sbt -mem 4096 -Pdocker-integration-tests "testOnly *PostgresKrbIntegrationSuite org.apache.spark.sql.jdbc.PostgresIntegrationSuite org.apache.spark.sql.jdbc.MySQLIntegrationSuite org.apache.spark.sql.jdbc.MsSqlServerIntegrationSuite org.apache.spark.sql.jdbc.MariaDBKrbIntegrationSuite *DB2KrbIntegrationSuite org.apache.spark.sql.jdbc.DB2IntegrationSuite org.apache.spark.sql.jdbc.OracleIntegrationSuite"
```

If so, I'm trying that now. Seems like a lib and some referenced docker images are out of date. I'm going to open another PR with those updates so I can play unit test roulette in a context I deem less time sensitive than this one.

tdg5 commented:
If I identified the correct integration tests, I can confirm they all pass on this branch when run manually. I'm out of time for the moment, so will have to wait to open the PR with the changes needed to make the docker integration tests work, but here's the diff so you can validate I didn't do anything crazy: https://github.com/tdg5/spark/commit/1d6fff5e313eb35541c2c04054b9cf9291c67962

HyukjinKwon commented:
@tdg5, thanks i think it's getting closer to be merged. but just to be sure, would you mind enabling GitHub Actions in your forked repository? See also https://github.com/apache/spark/pull/34745/checks?check_run_id=4569440425. Jenkins will be retired soon, and we're currently running GitHub Actions for tests (that actually covers more correctly).

I haven't followed the changes and issue here to be honest so I will leave this to @gaborgsomogyi.

HyukjinKwon commented on sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcConnectionProvider.scala:
```diff
@@ -53,12 +53,25 @@ abstract class JdbcConnectionProvider {
+   * @param driver  Java driver which initiates the connection
+   * @param options Driver options which initiates the connection
+   * @return True if the connection provider will need to modify the security configuration when
+   * initiating a connection with the given driver with the given options.
```
Let's also mention `@since 3.3.0`. Otherwise, looks pretty good.

HyukjinKwon replied:
I don't mind backporting it with, for example, `@since 3.2.1` or `@since 3.1.3` if you think this is a serious issue. I will leave this to you too @gaborgsomogyi.

gaborgsomogyi replied:
I think it worth a backport since it's a performance regression issue.

gaborgsomogyi commented:
The changeset looks good, except the `@since...` stuff. @tdg5 just to double check the last and most important thing do you have some performance numbers how much is it improved in unsecure case?

SparkQA commented:
**[Test build #146401 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/146401/testReport)** for PR 34745 at commit [`f7b6507`](https://github.com/apache/spark/commit/f7b65073da0a129cd3d01999f8978f0962b4b700).

tdg5 commented:
I added the missing `@since` with a value of `@since 3.1.3`. I also corrected the MIMA exception that I had added because I put it in the 3.1->3.2 bucket instead of the 3.2->3.3 bucket. Not sure if the `@since` changes that, but the MIMA test failed w/o the exception w/ the `@since` in place, so I imagine it is still required.

Not sure why the GitHub actions aren't working on my fork; they are enabled. I tweaked a couple of other things. Hopefully that helps.

Since I have to play unit test roulette due to the addition of `@since`, I added the library changes to make the docker integration tests work to this PR also. Let me know if that's a problem or if you have any thoughts on the changes.

tdg5 commented on external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/MariaDBKrbIntegrationSuite.scala:
```diff
@@ -25,9 +25,9 @@ import org.apache.spark.sql.execution.datasources.jdbc.connection.SecureConnecti
+ * To run this test suite for a specific version (e.g., mariadb:10.5.13):
  * {{{
- *   ENABLE_DOCKER_INTEGRATION_TESTS=1 MARIADB_DOCKER_IMAGE_NAME=mariadb:10.5.12
+ *   ENABLE_DOCKER_INTEGRATION_TESTS=1 MARIADB_DOCKER_IMAGE_NAME=mariadb:10.5.13
```
As far as I could tell the referenced image no longer exists in docker hub.

HyukjinKwon replied:
cc @sarutak

sarutak replied:
`10.5.12` is still available.
https://hub.docker.com/layers/mariadb/library/mariadb/10.5.12/images/sha256-a5ebad8f03ea9352ccf1ec3c84d91c2112ba86bbface1b57091f37aa7899d6c9?context=explore

But it's OK to upgrade.

sarutak replied:
@tdg5 If you'd like to change here, please change [this part](https://github.com/apache/spark/pull/34745/files#diff-cb12fa58307c2155fe2156d347d7a7917f1c755c34488a741dfb06653f83f364L41) too.

tdg5 replied:
I'm going to roll this back since it is not required; sorry for the confusion.

tdg5 commented on external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/MsSqlServerIntegrationSuite.scala:
```diff
@@ -26,9 +26,10 @@ import org.apache.spark.sql.internal.SQLConf
  * {{{
- *   ENABLE_DOCKER_INTEGRATION_TESTS=1 MSSQLSERVER_DOCKER_IMAGE_NAME=2019-CU13-ubuntu-20.04
+ *   ENABLE_DOCKER_INTEGRATION_TESTS=1
+ *   MSSQLSERVER_DOCKER_IMAGE_NAME=mcr.microsoft.com/mssql/server:2019-CU14-ubuntu-20.04
```
As far as I could tell the referenced image needs to be pulled from mcr.microsoft.com and the previously referenced image doesn't seem to exist anymore. https://hub.docker.com/_/microsoft-mssql-server

tdg5 replied:
Also, this example was missing the repository name, so I added that

HyukjinKwon replied:
cc @sarutak FYI

sarutak replied:
@tdg5 Thank you for fixing the example. Could you fix the example in `v2/MsSqlServerIntegrationSite` too?

sarutak replied:
Also, please change [this part](https://github.com/apache/spark/pull/34745/files#diff-cb12fa58307c2155fe2156d347d7a7917f1c755c34488a741dfb06653f83f364L40) too.

tdg5 replied:
hmm, this image also seems to exist. I'm going to roll that part back, but will fix the docs in another PR.

tdg5 commented on project/SparkBuild.scala:
```diff
@@ -600,7 +600,7 @@ object DockerIntegrationTests {
     dependencyOverrides += "com.google.guava" % "guava" % "18.0",
     resolvers += "DB2" at "https://app.camunda.com/nexus/content/repositories/public/",
-    libraryDependencies += "com.oracle" % "ojdbc6" % "11.2.0.1.0" from "https://app.camunda.com/nexus/content/repositories/public/com/oracle/ojdbc6/11.2.0.1.0/ojdbc6-11.2.0.1.0.jar" // scalastyle:ignore
+    libraryDependencies += "com.oracle" % "ojdbc6" % "11.2.0.3" from "https://app.camunda.com/nexus/content/repositories/public/com/oracle/ojdbc6/11.2.0.3/ojdbc6-11.2.0.3.jar" // scalastyle:ignore
```
The previously referenced version no longer exists in jfrog, so I updated to the closest similar version that was available. https://camunda.jfrog.io/ui/native/public/com/oracle/ojdbc6

HyukjinKwon replied:
cc @sarutak FYI

HyukjinKwon replied:
BTW, is it something we can backport up to branch-3.1? If not, let's separate these changes into a separate PR so that can be only merged to master or other branches.

sarutak replied:
Hmm, I'm confused about this part. Do we need this dependency setting here?
`pom.xml` has the setting for `ojdbc8`.

tdg5 replied:
I'm going to pull this into another PR for the sake of simplifying any backporting. I will update with a link shortly.

tdg5 commented:
@HyukjinKwon , as far as I can tell, it seems I succeeded in properly enabling GitHub actions for my fork 👍🏻 If you still see otherwise, please let me know.

@gaborgsomogyi , can you offer some thoughts on what kind of performance stats are you looking for? Since this is a threading/locking issue, the speed up of being able to open more connections simultaneously can be kind of unlimited depending on how many threads are allocated to doing that kind of work. I don't mind trying to come up with some numbers, but I'm not sure what makes sense to use as a baseline.

tdg5 commented:
Sorry, I think this is a dumb question, but I can't seem to find the answer with google. Is there a nightly snapshot build for 3.3 I can download somewhere?

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/50876/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/50876/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/50876/

SparkQA commented:
**[Test build #146408 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/146408/testReport)** for PR 34745 at commit [`4aac540`](https://github.com/apache/spark/commit/4aac54032ead4dd66f94de7726951aec91488a18).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/50883/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/50883/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/50883/

SparkQA commented:
**[Test build #146408 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/146408/testReport)** for PR 34745 at commit [`4aac540`](https://github.com/apache/spark/commit/4aac54032ead4dd66f94de7726951aec91488a18).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/146408/

SparkQA commented:
**[Test build #146401 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/146401/testReport)** for PR 34745 at commit [`f7b6507`](https://github.com/apache/spark/commit/f7b65073da0a129cd3d01999f8978f0962b4b700).
 * This patch **fails from timeout after a configured wait of `500m`**.
 * This patch merges cleanly.
 * This patch adds no public classes.

HyukjinKwon commented:
> Sorry, I think this is a dumb question, but I can't seem to find the answer with google. Is there a nightly snapshot build for 3.3 I can download somewhere?

It's not a dumb question :-) - I believe most of people don't know. Here you go: https://repository.apache.org/content/groups/snapshots/org/apache/spark

POM would have to have something like:

```xml
<repository>
  <id>apache-snapshot-repo</id>
  <name>Apache Snapshot Repository</name>
  <url>https://repository.apache.org/content/groups/snapshots</url>
  <releases>
    <enabled>false</enabled>
  </releases>
  <snapshots>
    <enabled>true</enabled>
  </snapshots>
</repository>
```

HyukjinKwon commented:
But I think current state of CI seems fine enough to go and merge.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/146401/

Commit:
[SPARK-37391][SQL]JdbcConnectionProvider tells if it modifies security context

Augments the JdbcConnectionProvider API such that a provider can
indicate that it will need to modify the global security configuration
when establishing a connection, and as such, if access to the global
security configuration should be synchronized to prevent races.

SparkQA commented:
**[Test build #146443 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/146443/testReport)** for PR 34745 at commit [`a7e9e2e`](https://github.com/apache/spark/commit/a7e9e2e8cbfa79b4900c685c740be8d8a394f0df).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/50919/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/50919/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/50919/

SparkQA commented:
**[Test build #146443 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/146443/testReport)** for PR 34745 at commit [`a7e9e2e`](https://github.com/apache/spark/commit/a7e9e2e8cbfa79b4900c685c740be8d8a394f0df).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/146443/

tdg5 commented:
The updated changes related to running docker integration tests can now be found over here: https://github.com/apache/spark/pull/34979

tdg5 commented:
I should have asked this earlier; how do backports work? Is there more y'all need from me in that regard? I've checked and the code in this PR doesn't cherry pick cleanly into either `branch-3.1` or `branch-3.2`.

HyukjinKwon commented:
> I should have asked this earlier; how do backports work? Is there more y'all need from me in that regard? I've checked and the code in this PR doesn't cherry pick cleanly into either `branch-3.1` or `branch-3.2`.

Yeah, it's all as you said. When committers merge, they usually manually resolve conflicts in their local if it's minor. If there's a major conflict, usually they ask the author to create another PR to `branch-3.1` or `branch-3.2`. While we can merge them as is first and you can create backporting PRs, it is often easier to split a PR to multiple PRs that can be backported and not backported.

HyukjinKwon commented:
Thanks @tdg5 - i like how this patch is thoroughly tested. this is a good example to follow :-).

tdg5 commented:
branch-3.1 PR is here: https://github.com/apache/spark/pull/34988

sarutak commented:
Merging to `master`. Thank you all !
